### PR TITLE
Re-writes storm auth logic and view controller in Swift

### PR DESCRIPTION
**Pull request by simonmitchell**
Fixes some bugs in LocalisationController
Moves StormLoginViewController and AuthenticationController to swift

### CODE REVIEW

File: ThunderCloud/AuthenticationController.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	public init() {
+		
+		requestController = TSCRequestController(baseAddress: "https://auth.cubeapis.com/v1.5")
+	}
+	
+	public func authenticateWith(username: String, password: String, completion: AuthenticationCompletion?) {
+		
+		requestController.post("authentication", bodyParams: ["username": username, "password": password]) { (response, error) in
+			
+			guard error == nil else {
```
Comment (ryanbourneuk) [inline ThunderCloud/AuthenticationController.swift]: This is exactly why I like `guard` so much, such a clean flow! Very nice indeed :)

File: ThunderCloud/LocalisationController.swift [hunk 2]
```diff
@@ -714,17 +719,20 @@ public class LocalisationController: NSObject {
 	
 	//MARK: - Logging in
 	//MARK: -
-	func askForLogin(completion: TSCStormLoginCompletion?) {
+	func askForLogin(completion: StormLoginViewController.LoginCompletion?) {
 		
 		let storyboard = UIStoryboard(name: "Login", bundle: Bundle.init(for: LocalisationController.self))
-		let loginViewController = storyboard.instantiateInitialViewController() as! TSCStormLoginViewController
+		let loginViewController = storyboard.instantiateInitialViewController() as! StormLoginViewController
+		loginViewController.loginReason = "Log in to your Storm account to start editing Localisations"
```
Comment (ryanbourneuk) [inline ThunderCloud/LocalisationController.swift]: Should `Localisations` be capitalised?

File: ThunderCloud/StormLoginViewController.swift [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+	
+	override func viewWillAppear(_ animated: Bool) {
+		
+		super.viewWillAppear(animated)
+		
+		UIView.animate(withDuration: 1.0, delay: 0.0, usingSpringWithDamping: 1.0, initialSpringVelocity: 0.0, options: [], animations: {
+			
+			self.backgroundView.alpha = 1.0
+			self.containerView.alpha = 1.0
+
```
Comment (ryanbourneuk) [inline ThunderCloud/StormLoginViewController.swift]: Is this blank line here needed?

File: ThunderCloud/StormLoginViewController.swift [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+		UIView.animate(withDuration: 1.0, delay: 0.0, usingSpringWithDamping: 1.0, initialSpringVelocity: 0.0, options: [], animations: {
+			
+			self.titleLabel.alpha = 0.0
+			self.reasonLabel.alpha = 0.0
+			self.passwordField.alpha = 0.0
+			self.usernameField.alpha = 0.0
+			self.loginButton.alpha = 0.0
+			
+			childView.alpha = 1.0
+
```
Comment (ryanbourneuk) [inline ThunderCloud/StormLoginViewController.swift]: Same for this one too, the one at the top is fine but it doesn't seem like other blocks have one at the end? What is preferred? :)

File: ThunderCloud/StormLoginViewController.swift [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+		
+		UIView.animate(withDuration: 1.0, delay: 0.0, usingSpringWithDamping: 1.0, initialSpringVelocity: 0.0, options: [], animations: {
+			
+			self.backgroundView.alpha = 1.0
+			self.containerView.alpha = 1.0
+			
+		}, completion: nil)
+	}
+	
+	private func updateBottomConstraint(with keyboardNotification: Notification) {
```
Comment (ryanbourneuk) [inline ThunderCloud/StormLoginViewController.swift]: Bit of documentation for this function could be good :)

File: ThunderCloud/LocalisationController.swift (hunk 2 above)
Comment (simonmitchell) [reply ThunderCloud/LocalisationController.swift]: Probably not, Copy pasta ;)

File: ThunderCloud/StormLoginViewController.swift (hunk 4 above)
Comment (simonmitchell) [reply ThunderCloud/StormLoginViewController.swift]: I tend to put a line break if there is another opening brace on the same line as a closing one:
```
if condition {

    print("Do something")

} else {
    
    print("Do something else")
}
```

**Comment by simonmitchell** [comment]
@ryanbourneuk this code is crashing somewhere, but can't figure out where... I'm looking into a solution now, but hold of for reviewing for the meantime!

File: ThunderCloud/AuthenticationController.swift [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	
+	/// A typealias for an authentication completion
+	public typealias AuthenticationCompletion = (_ authorization: Authorization?, _ error: Error?) -> Void
+	
+	private let requestController: TSCRequestController
+	
+	public init() {
+		
+		requestController = TSCRequestController(baseAddress: "https://auth.cubeapis.com/v1.5")
```
Comment (MattCheetham) [inline ThunderCloud/AuthenticationController.swift]: Should be `v1.6`

File: ThunderCloud/StormLoginViewController.swift [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+	/// The reason for requesting a storm login, will be displayed on the login UI
+	var loginReason: String?
+	
+	/// A view controller which will be shown inside the small white container view controller upon
+	/// sucessful login
+	var successViewController: UIViewController?
+	
+	@IBOutlet weak private var titleLabel: UILabel!
+	
+	@IBOutlet weak private var reasonLabel: UILabel!
```
Comment (MattCheetham) [inline ThunderCloud/StormLoginViewController.swift]: For the sake of clarity can you document up these properties?

File: ThunderCloud/StormLoginViewController.swift [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+		}
+		
+		if usernameField.isFirstResponder {
+			usernameField.resignFirstResponder()
+		}
+	}
+	
+	//MARK: - Action Handlers -
+	
+	@IBAction func handle1Password(_ sender: UIButton) {
```
Comment (MattCheetham) [inline ThunderCloud/StormLoginViewController.swift]: Document

File: ThunderCloud/StormLoginViewController.swift [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+			
+			if let username = loginDictionary[AppExtensionUsernameKey] as? String {
+				self.usernameField.text = username
+			}
+		}
+	}
+	
+	private var loginError: Error?
+	
+	@IBAction private func handleLogin(_ sender: TSCButton) {
```
Comment (MattCheetham) [inline ThunderCloud/StormLoginViewController.swift]: Document

File: ThunderCloud/StormLoginViewController.swift [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+			if strongSelf.isLoggedIn {
+				strongSelf.handleLoggedIn()
+			} else {
+				// Call completion for the sake of user maybe wanting to handle this situation
+				strongSelf.completion?(false, false, error)
+			}
+		}
+	}
+	
+	@IBAction private func handleDismissTap(_ sender: UITapGestureRecognizer) {
```
Comment (MattCheetham) [inline ThunderCloud/StormLoginViewController.swift]: Document

**Comment by MattCheetham** [comment]
Nice use of Encodable/Decodable. need to get on board with that!

File: ThunderCloud/AuthenticationController.swift (hunk 6 above)
Comment (simonmitchell) [reply ThunderCloud/AuthenticationController.swift]: I'm not sure if this is the case, I believe it's a different API with different versioning to the CMS content API

**Comment by simonmitchell** [comment]
@MattCheetham @ryanbourneuk changes made 😀

File: ThunderCloud/AuthenticationController.swift (hunk 6 above)
Comment (MattCheetham) [reply ThunderCloud/AuthenticationController.swift]: `https://auth.cubeapis.com/latest` says v1.6 is the latest. I remember Steve asking me why we were still using v1.5 a while back. Have you tried it and seen if it works?

File: ThunderCloud/StormLoginViewController.swift (hunk 8 above)
Comment (MattCheetham) [reply ThunderCloud/StormLoginViewController.swift]: Still needs documenting ;)

File: ThunderCloud/AuthenticationController.swift (hunk 6 above)
Comment (simonmitchell) [reply ThunderCloud/AuthenticationController.swift]: Will give it a go now! My bad, should've checked 👍🏻

**Comment by simonmitchell** [comment]
@MattCheetham @ryanbourneuk  3rd time lucky 😝

**Comment by simonmitchell** [comment]
@MattCheetham @ryanbourneuk Just a heads up made some important fixes to this (4th time lucky) on the latest commit when you come to review this again 👍

**Pull request merged**